File: richdoc/richdoc-cli/src/commands/new.ts
import { existsSync } from "node:fs";
import { copyFile, mkdir } from "node:fs/promises";
import { dirname, resolve } from "node:path";
import { defineCommand } from "citty";
import { ASSET_FILES } from "../lib/assets.ts";
import { jsonError, jsonOk } from "../lib/output.ts";
import {
	listTemplates,
	templateExists,
	templatePath,
} from "../lib/templates.ts";

const DEFAULT_TEMPLATE = "plan";

export const newCommand = defineCommand({
	meta: {
		name: "new",
		description:
			"Scaffold a new richdoc .html file from a template. Use `richdoc init` next to it to drop the CSS/JS assets.",
	},
	args: {
		output: {
			type: "positional",
			description: "Output .html path (e.g. docs/plan.html)",
			required: true,
		},
		template: {
			type: "string",
			alias: "t",
			description: `Template to use (default: ${DEFAULT_TEMPLATE})`,
			default: DEFAULT_TEMPLATE,
		},
		force: {
			type: "boolean",
			alias: "f",
			description: "Overwrite the output file if it exists",
			default: false,
		},
	},
	async run({ args }) {
		const tplName = args.template || DEFAULT_TEMPLATE;
		const available = listTemplates();
		if (!templateExists(tplName)) {
			jsonError(
				`Unknown template '${tplName}'.`,
				"TEMPLATE_NOT_FOUND",
				`Available templates: ${available.join(", ")}`,
				{ available },
			);
		}

		const output = resolve(args.output);
		if (!output.toLowerCase().endsWith(".html")) {
			jsonError(
				`Output path must end with .html (got '${output}').`,
				"INVALID_PARAMS",
			);
		}
		if (existsSync(output) && !args.force) {
			jsonError(
				`Output file already exists: ${output}`,
				"FILE_EXISTS",
				"Re-run with --force to overwrite.",
				{ file: output },
			);
		}

		try {
			await mkdir(dirname(output), { recursive: true });
			await copyFile(templatePath(tplName), output);
		} catch (err) {
			const reason = err instanceof Error ? err.message : "unknown error";
			jsonError(`Could not write output: ${reason}`, "OUTPUT_ERROR");
		}

		// Check whether the assets are present next to the output. If not,
		// surface a hint — the linked `./richdoc.css` / `./richdoc.js` won't
		// resolve until `richdoc init` runs in the same directory.
		const dir = dirname(output);
		const missing = ASSET_FILES.filter((f) => !existsSync(resolve(dir, f)));

		jsonOk({
			file: output,
			template: tplName,
			assets_needed: missing,
			hint: missing.length
				? `Run \`richdoc init ${dir}\` to drop the CSS/JS assets next to this file.`
				: undefined,
		});
	},
});

File: richdoc/richdoc-cli/src/richdoc_cli/commands/new_.py
"""`richdoc new` — scaffold a new .html document from a template."""

from __future__ import annotations

import shutil
from pathlib import Path

import click

from ..assets import ASSET_FILES
from ..output import json_error, json_ok
from ..templates import list_templates, template_exists, template_path

DEFAULT_TEMPLATE = "plan"


@click.command("new")
@click.argument(
    "output",
    type=click.Path(dir_okay=False, path_type=Path),
)
@click.option(
    "-t",
    "--template",
    "template",
    default=DEFAULT_TEMPLATE,
    show_default=True,
    help="Template to use.",
)
@click.option(
    "-f",
    "--force",
    is_flag=True,
    help="Overwrite the output file if it exists.",
)
def cmd(output: Path, template: str, force: bool) -> None:
    """Scaffold a new richdoc .html file from a template."""
    available = list_templates()
    if not template_exists(template):
        json_error(
            f"Unknown template '{template}'.",
            code="TEMPLATE_NOT_FOUND",
            hint=f"Available templates: {', '.join(available) if available else '(none)'}",
            available=available,
        )

    out_path = output.resolve()
    if out_path.suffix.lower() != ".html":
        json_error(
            f"Output path must end with .html (got '{out_path}').",
            code="INVALID_PARAMS",
        )

    if out_path.exists() and not force:
        json_error(
            f"Output file already exists: {out_path}",
            code="FILE_EXISTS",
            hint="Re-run with --force to overwrite.",
            file=str(out_path),
        )

    try:
        out_path.parent.mkdir(parents=True, exist_ok=True)
        shutil.copyfile(template_path(template), out_path)
    except OSError as exc:
        json_error(
            f"Could not write output: {exc}",
            code="OUTPUT_ERROR",
        )

    # Whether the linked assets are present next to the output file.
    parent = out_path.parent
    missing = [f for f in ASSET_FILES if not (parent / f).is_file()]
    hint = (
        f"Run `richdoc init {parent}` to drop the CSS/JS assets next to this file."
        if missing
        else None
    )
    json_ok(
        file=str(out_path),
        template=template,
        assets_needed=missing,
        **({"hint": hint} if hint else {}),
    )
